### PR TITLE
REF: implement shift_bday

### PR DESCRIPTION
**Pull request by jbrockmendel**
Avoids depending on DatetimeArray/PeriodArray methods, also avoids a couple of array allocations

```
In [2]: dti = pd.date_range("2016-01-01", periods=10**5, freq="S")                                                                     

In [3]: off = pd.offsets.BDay()                                                                                                        

In [4]: %timeit dti + off                                                                                                              
24.1 ms ± 202 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  # <-- master
20.1 ms ± 664 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  # <-- PR
```

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -3991,6 +3974,51 @@ def shift_months(const int64_t[:] dtindex, int months, object day=None):
     return np.asarray(out)
 
 
+cdef ndarray[int64_t] shift_bdays(const int64_t[:] i8other, int periods):
+    cdef:
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:3978]: can you doc-string args

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -3991,6 +3974,51 @@ def shift_months(const int64_t[:] dtindex, int months, object day=None):
...
+        int64_t val, res
+        int wday, nadj, days
+        npy_datetimestruct dts
+
+    for i in range(n):
+        val = i8other[i]
+        if val == NPY_NAT:
+            result[i] = NPY_NAT
+        else:
+            # The rest of this is effectively a copy of BusinessDay.apply
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:3990]: TODO here to consolidate?

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -3991,6 +3974,51 @@ def shift_months(const int64_t[:] dtindex, int months, object day=None):
     return np.asarray(out)
 
 
+cdef ndarray[int64_t] shift_bdays(const int64_t[:] i8other, int periods):
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx:3977]: this can be nogil?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/offsets.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:3990]: not really feasible; different types.

File: pandas/_libs/tslibs/offsets.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx:3977]: thats giving me compile-time errors, may revisit, pretty low-priority

**Review by jreback — APPROVED** [review]

**Pull request merged**